### ISSUE TITLE
dictionary of metrics are not moved to device (GPU)

### ISSUE THREAD

**Issue by astrocyted** [Lightning-AI/lightning#17624]
### Bug description

when I define metrics (from torchmetrics) in the __init__ of my lightning module as elements ina dictionary, they are not moved to cuda/device and will stay on cpu.
where as if I define them as separate variables (not inside a dict) they are successfully moved to device.
Is that a bug ?

### What version are you seeing the problem on?

v1.9

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow):
#- PyTorch Lightning Version (e.g., 1.5.0):
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 2.0):
#- Python version (e.g., 3.9):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):
```

</details>

**Comment by astrocyted** [comment]
Issue can be deleted. just found the answer on torchmetrics docs : 

> However, when properly defined inside a [Module](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module) or LightningModule the metric will be automatically moved to the same device as the module when using .to(device). Being properly defined means that the metric is correctly identified as a child module of the model (check .children() attribute of the model). Therefore, metrics cannot be placed in native python list and dict, as they will not be correctly identified as child modules. Instead of list use [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html#torch.nn.ModuleList) and instead of dict use [ModuleDict](https://pytorch.org/docs/stable/generated/torch.nn.ModuleDict.html#torch.nn.ModuleDict). Furthermore, when working with multiple metrics the native [MetricCollection](https://torchmetrics.readthedocs.io/en/stable/pages/overview.html#metriccollection) module can also be used to wrap multiple metrics.

**Comment by awaelchli** [comment]
@astrocyted Great that you found the answer. I can confirm it is the expected behavior. Closing :)
